src/ralph/models/edx/converters/xapi/server.py

```diff
@@ -0,0 +1,75 @@
+"""Server event xAPI Converter"""
+
+from ralph.models.converter import GetFromField
+from ralph.models.xapi.constants import (
+    XAPI_EXTENSION_ACCEPT_LANGUAGE,
+    XAPI_EXTENSION_AGENT,
+    XAPI_EXTENSION_COURSE_ID,
+    XAPI_EXTENSION_COURSE_USER_TAGS,
+    XAPI_EXTENSION_HOST,
+    XAPI_EXTENSION_IP,
+    XAPI_EXTENSION_ORG_ID,
+    XAPI_EXTENSION_PATH,
+    XAPI_EXTENSION_REQUEST,
+)
+from ralph.models.xapi.page_viewed import PageViewed
+
+from .base import BaseXapiConverter
+
+
+class ServerEventToXapi(BaseXapiConverter):
+    """Converts a common edX server event to xAPI.
+
+    Example Statement: John viewed https://www.fun-mooc.fr/ Web page.
+
+    WARNING: Duplicate xAPI values:
+        - object.id == context.extensions.XAPI_EXTENSION_PATH.
+
+    WARNING: The converter does not include the following edX fields:
+        - username: As it is similar to the `context.user_id` field.
+        - referer: As it may contain sensitive data.
+        - event_source: As it is replaced by xAPIs `object.definition`.
+
+    WARNING: The converter may not include the following edX fields:
+        - ip: When `ip` is an empty string.
+        - agent: When `agent` is an empty string.
+        - accept_language: When `accept_language` is an empty string.
+        - context.course_user_tags: When `course_user_tags` is an empty dictionary.
+        - context.org_id: When `org_id` is an empty string.
+        - context.course_id: When `course_id` is an empty string.
+    """
+
+    __model__ = PageViewed
+
+    def get_conversion_dictionary(self):
+        """Returns a conversion dictionary used for conversion."""
+
+        return {
+            "actor": {
+                "account": {
+                    "name": GetFromField("context__user_id"),
+                    "homePage": self.platform,
+                }
+            },
+            "context": {
+                "extensions": {
+                    XAPI_EXTENSION_ACCEPT_LANGUAGE: GetFromField("accept_language"),
+                    XAPI_EXTENSION_AGENT: GetFromField("agent"),
+                    XAPI_EXTENSION_COURSE_ID: GetFromField("context__course_id"),
+                    XAPI_EXTENSION_COURSE_USER_TAGS: GetFromField(
+                        "context__course_user_tags"
+                    ),
+                    XAPI_EXTENSION_HOST: GetFromField("host"),
+                    XAPI_EXTENSION_IP: GetFromField("ip"),
+                    XAPI_EXTENSION_ORG_ID: GetFromField("context__org_id"),
+                    XAPI_EXTENSION_PATH: GetFromField("context__path"),
+                    XAPI_EXTENSION_REQUEST: GetFromField("event"),
+                },
+            },
+            "object": {
+                "id": GetFromField(
+                    "event_type", lambda event_type: self.platform + event_type
+                )
```

Comment:
I tried to replace `GetFromField` with `LazyModelField` without success for now.
`platform` is not in the edx event but provided with an cli argument.
In pydantic it's possible to define [private model attributes](https://pydantic-docs.helpmanual.io/usage/models/#private-model-attributes) but it seems there is no way to access them from a validator function ? Any Ideas ?

Reply:
Just to understand better when using `LazyModelField`, you don't apply the transformation `x : x+ self.platform` but the [path parameter](https://github.com/openfun/ralph/blob/master/src/ralph/models/selector.py#L24) is directly `x + self.platform` ?

Reply:
Yes, `LazyModelField` don't do any transformations it helps just to retrieve a value from the event.

`GetFromField`on the other hand  retrieves the value from the event and applies a transformation.

If I understood it right, we want to pass the value (of `event_type`) "as it is" in the source event to the pydantic xAPI model and then let the model do the transformation part (x + self.platform).

But I don't find a way to pass `self.platform` to the pydantic xAPI model and access it's value form the validator to do the transformation (Without using global or static variables).